### PR TITLE
docs: new monitoring and metrics docs (DOCS-3693)

### PR DESCRIPTION
**Pull request by MichaelDrogalis**
### Description 

The current monitoring documentation suggests an approach that is defunct. This PR refreshes our recommendations on this topic.

@JimGalasyn One consideration that we may want to make is mirroring this page in our enterprise docs, but also describing how to set up with on a tar/rpm package, etc. In that case, all you need to do is export a `JMX_PORT` variable without supplying more configuration.

### CODE REVIEW

File: docs/operate-and-deploy/monitoring-and-metrics.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
+# Monitoring and monitoring
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring-and-metrics.md:1]:
```suggestion
# Monitoring and metrics
```

File: docs/operate-and-deploy/monitoring-and-metrics.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+## Context
+
+ksqlDB publishes metrics via JMX ([Java Management
+Extensions](https://www.oracle.com/java/technologies/javase/javamanagement.html))
+which help you monitor what is happening inside of ksqlDB's servers. A
+comprehensive set of metrics are published, including number of rows
+processed, throughput, throughput distribution, and error rate.
+
+## Enabling monitoring
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring-and-metrics.md:11]:
```suggestion
## Enable monitoring
```

File: docs/operate-and-deploy/monitoring-and-metrics.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+You must explictly enable monitoring on each ksqlDB server. To enable
+it in a Docker-based deployment, export an environment variable named
+`KSQL_JMX_OPTS` with your JMX configuration and expose the port that
+JMX will communicate over.
+
+To illustrate how that should look, take a look at this Docker Compose
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring-and-metrics.md:18]:
```suggestion
The following Docker Compose
```

File: docs/operate-and-deploy/monitoring-and-metrics.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+snippet. This is an example of how ksqlDB server can be configured for
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring-and-metrics.md]:
```suggestion
example shows how you can configure monitoring for ksqlDB server.
```

File: docs/operate-and-deploy/monitoring-and-metrics.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+monitoring. The surrounding components, like the broker and CLI, are
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring-and-metrics.md]:
```suggestion
The surrounding components, like the broker and CLI, are
```

File: docs/operate-and-deploy/monitoring-and-metrics.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+- The environment variable `KSQL_JMX_OPTS` is supplied to the server
+  with various arguments. The `>` character lets you write a
+  multi-line string in Yaml, which makes this long argument easier to
+  read. The advertised hostname, port, and security settings are
+  configured. JMX has a wide range of [configuration
+  options](https://docs.oracle.com/javase/8/docs/technotes/guides/management/agent.html),
+  and you can set these however you like.
+
+- Port `1099` is exposed, which corresponds to the JMX port set in the
+  `KSQL_JMX_OPTS` configuration. This allows remote monitoring tools
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring-and-metrics.md]:
```suggestion
  `KSQL_JMX_OPTS` configuration. This enables remote monitoring tools
```

File: docs/operate-and-deploy/monitoring-and-metrics.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  to communicate into ksqlDB's process.
+
+## Verifying your monitoring setup
+
+An easy way to check that ksqlDB is properly emitting metrics is by
+using `jconsole`. JConsole is a graphical monitoring tool to monitor
+the JVM, and it ships with by default with Oracle JDK installations.
+
+On your host machine, simply run the command:
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring-and-metrics.md]:
```suggestion
On your host machine, run the command:
```

File: docs/operate-and-deploy/monitoring-and-metrics.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+```
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring-and-metrics.md]:
```suggestion
```bash
```

File: docs/operate-and-deploy/monitoring-and-metrics.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+jconsole
+```
+
+You will be prompted for a host and port. If you used the example
+configuration, `localhost:1099` will establish the connection. You
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring-and-metrics.md]:
```suggestion
configuration, `localhost:1099` establishes the connection. You
```

**Comment by MichaelDrogalis** [comment]
Thanks @JimGalasyn, more to come.

File: docs/reference/metrics.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
+# Metrics
+
+TODO: intro text.
+
+TODO: move - Because ksqlDB persistent queries directly compile into {{
+site.kstreams }} topologies, many ksqlDB metrics correspond to [Kafka
+Streams metrics](https://docs.confluent.io/current/streams/monitoring.html).
```
Comment (colinhicks) [inline docs/reference/metrics.md:7]: I realize this part is within a TODO, but I am curious if this means that ksqlDB server directly provides the `kafka.streams:*` MBeans listed on the linked page. For instance, should I expect thread and task metrics to be provided directly by ksqlDB's JMX endpoint, or would I have to gather them from JMX on the broker?

Comment (MichaelDrogalis) [reply docs/reference/metrics.md:7]: I'm going to dig in and find out. :)

Comment (MichaelDrogalis) [reply docs/reference/metrics.md]: > I am curious if this means that ksqlDB server directly provides the kafka.streams:*

It does.

**Comment by MichaelDrogalis** [comment]
I'm done cleaning up this section, but I have a few questions whose answers would help improve the docs further. Tagging people who I think would know:

1. Why are all the Vert.x metrics at the top level? All of our other metrics are tucked into a namespace, but the Vert.x ones (20+ of them) are hanging out at the top-level. Is that intentional? (e.g. `"_confluent-ksql-default_.http.servers.0.0.0.0:8088.trace-requests"`) @purplefox @vcrfxia 

2. Many of our metrics, like error rate (`_confluent-ksql-default_error-rate`) and pull query request rate (`pull-query-requests-rate`) describe behavior over time, but what is the rate against? Is it instances in the last 15 minutes? Of all time? @AlanConfluent @agavra 

3. Why do you need to explicitly enable pull query metrics? @AlanConfluent 

4. Can you/should you switch the prefix of `_confluent-ksql-default` on many of these metrics? Maybe @rodesai ?

5. Our "engine" metrics have a metric for total bytes consumed (`_confluent-ksql-default_bytes-consumed-total`) but not bytes produced. Intentional, or should I file an issue to add it? @agavra

**Comment by AlanConfluent** [comment]
> 2. Many of our metrics, like error rate (`_confluent-ksql-default_error-rate`) and pull query request rate (`pull-query-requests-rate`) describe behavior over time, but what is the rate against? Is it instances in the last 15 minutes? Of all time? @AlanConfluent @agavra
`pull-query-requests-rate` is requests per second, across all instances in a cluster, though it may be averaged over some window.  I assume `_confluent-ksql-default_error-rate` is similar, though someone else can speak to that.

> 3. ~Why do you need to explicitly enable pull query metrics? @AlanConfluent~ #5821
I agree.  It may as well be enabled by default.

**Comment by vcrfxia** [comment]
> Why are all the Vert.x metrics at the top level? All of our other metrics are tucked into a namespace, but the Vert.x ones (20+ of them) are hanging out at the top-level. Is that intentional? (e.g. "_confluent-ksql-default_.http.servers.0.0.0.0:8088.trace-requests") @purplefox @vcrfxia

By "namespace" do you mean the domain portion of the JMX MBean name? If so, it's being set to `io.confluent.ksql.metrics` for the Vert.x server metrics: https://github.com/confluentinc/ksql/blob/1ab37ba7c5d89cafa3ff46585af51ded3fb048bb/ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java#L1006

**Comment by MichaelDrogalis** [comment]
@vcrfxia I'm not certain of the right JMX terminology, but here's a screenshot(s) to show what I mean. I had to break it into two since it scrolls:

Screen Shot 2020-07-20 at 12 56 09 PM
Screen Shot 2020-07-20 at 12 56 27 PM

To contrast, you can see how the engine metrics tuck neatly under a single category. Anyway, we don't need to solve this here and can make a new issue if you think it's something to be fixed, I just wanted to verify that this is the way it ought to be right now.

**Comment by vcrfxia** [comment]
> To contrast, you can see how the engine metrics tuck neatly under a single category.

From the screenshot it looks like the ksqlDB engine metrics are a single MBean with different attributes for the different metrics, whereas the Vert.x server metrics are multiple MBeans for the different metrics. I'm not aware of a way to update the Vert.x server metrics to be structured like the ksqlDB engine metrics are, since if you click into the individual Vert.x server metrics you'll see that they also have attributes representing different stats for each metric. Don't let me dissuade you from filing a GitHub issue, but I don't think this is easily changed.

**Comment by MichaelDrogalis** [comment]
Understood, thanks @vcrfxia.

**Comment by MichaelDrogalis** [comment]
@JimGalasyn Im good with shipping this now if you are.

File: docs/reference/metrics.md [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+
+**Pull query latency 99th percentile**
+
+`pull-query-requests-latency-distribution-99`
+
+Latency distribution of the 99th percentile.
+
+## Command runner
+
+Metrics that describe the health of the CommandRunner thread, which
```
Comment (JimGalasyn) [inline docs/reference/metrics.md:319]:
```suggestion
Metrics that describe the health of the `CommandRunner` thread, which
```

File: docs/reference/metrics.md [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+## Pull queries
+
+Metrics that describe the activity of pull queries on each server.
+
+```
+io.confluent.ksql.metrics:type=_confluent-ksql-default_pull-query
+```
+
+!!! info
+    Note that pull query metrics must explicitly be enabled by setting
```
Comment (JimGalasyn) [inline docs/reference/metrics.md:222]:
```suggestion
    Pull query metrics must be enabled explicitly by setting
```

File: docs/reference/metrics.md [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+
+`_confluent-ksql-default_messages-produced-per-sec`
+
+The number of messages produced per second across all queries.
+
+**Error rate**
+
+`_confluent-ksql-default_error-rate`
+
+The number of messages which were consumed but not processed. Messages may not be processed if, for instance, the message contents could not be deserialized due to an incompatible schema. Alternately, a consumed messages may not have been produced, hence being effectively dropped. Such messages would also be counted toward the error rate.
```
Comment (JimGalasyn) [inline docs/reference/metrics.md:117]:
```suggestion
The number of messages that were consumed but not processed. Messages may not be processed if, for instance, the message contents could not be deserialized due to an incompatible schema. Alternately, a consumed message may not have been produced, hence being effectively dropped. Such messages would also be counted toward the error rate.
```

Comment (JimGalasyn) [reply docs/reference/metrics.md:117]: Consider line breaks at col 80.

File: docs/operate-and-deploy/monitoring.md [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+ksqlDB publishes metrics via JMX ([Java Management
+Extensions](https://www.oracle.com/java/technologies/javase/javamanagement.html))
+which help you monitor what is happening inside of ksqlDB's servers. A
+comprehensive [set of metrics](../reference/metrics.md) are published,
+including number of rows processed, throughput,
+throughput distribution, and error rate.
+
+## Enable monitoring
+
+You must explictly enable monitoring on each ksqlDB server. To enable
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring.md:14]:
```suggestion
You must enable monitoring explicitly on each ksqlDB server. To enable
```

File: docs/operate-and-deploy/monitoring.md [hunk 15]
```diff
@@ -0,0 +1,81 @@
+# Monitoring
+
+## Context
+
+ksqlDB publishes metrics via JMX ([Java Management
+Extensions](https://www.oracle.com/java/technologies/javase/javamanagement.html))
+which help you monitor what is happening inside of ksqlDB's servers. A
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring.md:7]:
```suggestion
which help you monitor what is happening inside of ksqlDB's servers. For a
```

File: docs/operate-and-deploy/monitoring.md [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+comprehensive [set of metrics](../reference/metrics.md) are published,
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring.md:8]:
```suggestion
comprehensive list of metrics, including number of rows processed, throughput,
```

File: docs/operate-and-deploy/monitoring.md [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+including number of rows processed, throughput,
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring.md:9]: throughput distribution, and error rate, see [Metrics](../reference/metrics.md).

File: docs/operate-and-deploy/monitoring.md [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+throughput distribution, and error rate.
```
Comment (JimGalasyn) [inline docs/operate-and-deploy/monitoring.md:10]:
```suggestion

```

**Comment by MichaelDrogalis** [comment]
Whoops, switched it out of draft mode.

**Comment by MichaelDrogalis** [comment]
Hooray!

**Comment by JimGalasyn** [comment]
@MichaelDrogalis should I cherry-pick to the live branch (0.10.1-ksqldb)?

**Comment by MichaelDrogalis** [comment]
Yes.

**Pull request merged**